# Prepare for applications
library(ape) 

system('mkdir -p app/www/')
system('cd app/www/ ; ln -s ../../database database')
system('cd app/www/ ; ln -s ../../data data')


## Tree preparation
list_files = list.files("data/dnds_phylo",recursive = T,pattern = "root.nwk" ,full.names = T)

phylo=data.frame()
for (file in list_files){
  print(file)
  tree = read.tree(file)
  set = str_split_1(str_replace(file,"data/dnds_phylo/",""),"/")[1]
  nb_genes = ""
  if (file.exists(paste("data/dnds_phylo/",set,"/readme",sep=""))){
    readme = read.table(paste("data/dnds_phylo/",set,"/readme",sep=""))
    rownames(readme) = readme$V1
    nb_genes = readme[3,"V2"]
  } 
  nb_species = length(tree$tip.label)
  
  phylo = rbind(phylo,
                data.frame(
                  name = file,
                  description = paste(set,nb_species,"species",nb_genes,"genes")
                ))
}

phylo = phylo[sapply(phylo$name,function(x) str_count(x, "/") <= 4),]

write.table(phylo , paste("app/www/phylogenetic_trees_description.tab",sep=""),quote=F,row.names = F,col.names = T,sep="\t")



## Informations table
options(stringsAsFactors = F, scipen = 999)

library(rgbif)
library(ape)
library(ggpubr)
library(ggrepel)
library(readxl)
library(seqinr)
library(stringr)
library(dplyr)
library(tidyr)

add_charac <- function(data_summary,label,description,value){
  data_summary = rbind(data_summary,data.frame(
    label = label,
    description = description,
    value=value
  ))
  return(data_summary)
}



get_rsquared_slope = function(prop.quantile = 0.2,Xaxis,Yaxis){
  # print(Xaxis)
  # print(Yaxis)
  if (length(Xaxis) != 0){ # Pas de vecteur vide
    quantile = unique(quantile(Xaxis, probs = seq(0, 1,prop.quantile),na.rm=T))
    if (length(quantile) > 2){ # au moins un intervalle
      intervalle = cut(Xaxis, quantile,include.lowest = T,include.higher=T)
      X = tapply(Xaxis, intervalle,  function(x) mean(x, na.rm=T))
      Y = tapply(Yaxis, intervalle, function(x) mean(x, na.rm=T))
      # print(Y)
      # print(X)
      formule = lm( formula = Y ~ X)
      R.signed = summary(formule)$coefficients["X","Estimate"] / abs(summary(formule)$coefficients["X","Estimate"])*round(summary(formule)$adj.r.squared, 3)
      slope = summary(formule)$coefficients["X","Estimate"]
      
      return(c(R.signed,slope))
    } else {
      return(c(NA,NA))
    }} else {
      return(c(NA,NA))
    }
}



data1 = read.delim("www/data/data1.tab")
rownames(data1) = data1$species
data1$path_db = paste(data1$species,"_NCBI.taxid",data1$NCBI.taxid,"/",data1$assembly_accession,sep="")

all_dt = data.frame()
for (species in data1$species ){print(species)
  pathData = "/home/fbenitiere/data/Projet-SplicedVariants/"
  
  gff_path = paste(pathData , "Annotations/",species,"/data_source/annotation.gff",sep="")
  gc_table_path = paste(pathData, "Annotations/",species,"/GC_content.tab",sep="")
  by_gene_analysis_path = paste("www/database/Transcriptomic/",data1[species,]$path_db,"/by_gene_analysis.tab.gz",sep="")
  by_intron_analysis_path = paste("www/database/Transcriptomic/",data1[species,]$path_db,"/by_intron_analysis.tab.gz",sep="")
  prot_path = paste(pathData , "Annotations/",species,"/data_source/protein.faa",sep="")
  
  data_summary = data.frame()
  data_summary = add_charac(data_summary,'species',"",species)
  
  con <- file(gff_path,"r")
  first_line <- readLines(con,n=10)
  close(con)
  print(first_line[5])
  genome_assembly = first_line[5]
  genome_assembly = str_replace(genome_assembly,"#!genome-build-accession NCBI_Assembly:","")
  data_summary = add_charac(data_summary,'genome_assembly',"",genome_assembly)
  
  
  data_summary = add_charac(data_summary,'clade;qual',"", data1[species,]$clade)
  data_summary = add_charac(data_summary,'lifespan_days;quant',"", data1[species,]$max_lifespan_days)
  data_summary = add_charac(data_summary,'length_cm;quant',"", data1[species,]$max_length_cm)
  data_summary = add_charac(data_summary,'weight_cm;quant',"", data1[species,]$max_weight_kg)
  
  key = name_backbone(name=str_replace(species,"_"," "),rank="species")$usageKey
  RGBIF_count = occ_search(key,limit=0)$meta$count
  
  data_summary = add_charac(data_summary,'RGBIF_observation;quant',"",RGBIF_count)
  
  
  genome_character = read.table(gc_table_path,header=T)
  genome_size_var =  sum(genome_character[genome_character$genome_character %in% c("A","T","G","C"), ]$Freq) / 1000000
  GC_content_proportion_var = round( sum(genome_character[genome_character$genome_character %in% c("G","C"), ]$Freq) /
                                       sum(genome_character[genome_character$genome_character %in% c("A","T","G","C"), ]$Freq) ,2)
  
  data_summary = add_charac(data_summary,'genome_size;quant',"Mb",genome_size_var)
  data_summary = add_charac(data_summary,'GC_content_proportion;quant',"",GC_content_proportion_var)
  data_summary = add_charac(data_summary,'No_prot_annot;quant',"",length(read.fasta(prot_path)))
  
  #### Data_set
  if (file.exists(by_gene_analysis_path)){
    by_gene =  read.delim(by_gene_analysis_path , header=T , sep="\t",comment.char = "#")
    by_gene = by_gene[by_gene$type == "gene" & grepl("gene_biotype=protein_coding" , by_gene$attributes),]
    rownames(by_gene) = by_gene$gene_id
    by_gene = by_gene[by_gene$type == "gene",] # FILTRE PSEUDOGENE
    
    by_intron =  read.delim(by_intron_analysis_path , header=T , sep="\t",comment.char = "#")
    by_intron = by_intron[by_intron$gene_id %in% by_gene$gene_id,] # FILTRE PSEUDOGENE
    by_intron$median_fpkm = by_gene[by_intron$gene_id,]$median_fpkm
    
    for (busco_group in c("metazoa","embryophyta","eukaryota","None")){ print(busco_group)
      can_analyse = T
      if ( busco_group != "None" ){
        if (file.exists(paste("www/database/BUSCO_annotations/",data1[species,]$path_db,"/busco_to_gene_id_",busco_group,sep=""))){
          busco_to_gene = read.delim(paste("www/database/BUSCO_annotations/",data1[species,]$path_db,"/busco_to_gene_id_",busco_group,sep=""))
          
          by_gene$busco_id = by_gene$gene_id %in% busco_to_gene$gene_id
          by_gene_selected = by_gene[by_gene$busco_id,]
          
          data_summary = add_charac(data_summary,paste("median_coverage_exon;buscodataset_",busco_group,";quant",sep=""),"",median(by_gene_selected$exon_coverage,na.rm = T))
          
          busco_to_gene = busco_to_gene[!(duplicated(busco_to_gene$busco_id,fromLast = FALSE) | duplicated(busco_to_gene$busco_id,fromLast = TRUE)) &
                                          !(duplicated(busco_to_gene$gene_id,fromLast = FALSE) | duplicated(busco_to_gene$gene_id,fromLast = TRUE)) ,]
          
          rownames(busco_to_gene) = busco_to_gene$gene_id
          by_gene$busco_id = by_gene$gene_id %in% busco_to_gene$gene_id
          by_intron$busco_id = by_intron$gene_id %in% busco_to_gene$gene_id
          
          by_gene_selected = by_gene[by_gene$busco_id,]
          by_intron_selected = by_intron[by_intron$busco_id & by_intron$intron_class == "major" & by_intron$into_cds == "True",]
          print(nrow(by_intron_selected))
          
        } else { can_analyse = F }
      } else {
        by_gene_selected = by_gene
        by_intron_selected = by_intron[ by_intron$intron_class == "major" & by_intron$into_cds == "True",]
        data_summary = add_charac(data_summary,paste("median_coverage_exon;buscodataset_",busco_group,";quant",sep=""),"",median(by_gene_selected$exon_coverage,na.rm = T))
      }
      
      if (can_analyse){
        data_summary = add_charac(data_summary,paste("No_genes;buscodataset_",busco_group,";quant",sep=""),"",nrow(by_gene_selected))
        data_summary = add_charac(data_summary,paste("No_introns;buscodataset_",busco_group,";quant",sep=""),"",nrow(by_intron_selected))
        data_summary = add_charac(data_summary,paste("median_length_introns;buscodataset_",busco_group,";quant",sep=""),"",median(abs(by_intron_selected$splice3-by_intron_selected$splice5)))
        data_summary = add_charac(data_summary,paste("No_multi_exonic_genes;buscodataset_",busco_group,";quant",sep=""),"",length(unique(by_intron_selected$gene_id)))
        data_summary = add_charac(data_summary,paste("No_intron_per_gene;buscodataset_",busco_group,";quant",sep=""),"",nrow(by_intron_selected) / length(unique(by_intron_selected$gene_id)))
        data_summary = add_charac(data_summary,paste("prop_intron_annotated;buscodataset_",busco_group,";quant",sep=""),"",sum(by_intron_selected$isAnnotated) / nrow(by_intron_selected))
        
        
        data_summary = add_charac(data_summary,paste("median_gene_fpkm;buscodataset_",busco_group,";quant",sep=""),"",median(by_gene_selected$median_fpkm,na.rm = T))
        
        for (svr_class in c("all" , "high_SV" , "low_SV")){
          by_intron_selected_svr = by_intron_selected
          if (svr_class == "all"){print("all")} else if ( svr_class == "high_SV" ){
            print("high_SV")
            by_intron_selected_svr = by_intron_selected[  by_intron_selected$splice_variant_rate >= 0.05 ,]
            print(nrow(by_intron_selected_svr))
          } else if (svr_class == "low_SV") {
            print("low_SV")
            by_intron_selected_svr = by_intron_selected[  by_intron_selected$splice_variant_rate < 0.05 ,]
            print(nrow(by_intron_selected_svr))
          }
          
          data_summary = add_charac(data_summary,paste("average_svr;svr_class_",svr_class,";buscodataset_",busco_group,";quant",sep=""),"",mean(by_intron_selected_svr$splice_variant_rate))
          data_summary = add_charac(data_summary,paste("average_nsvr;svr_class_",svr_class,";buscodataset_",busco_group,";quant",sep=""),"",mean(by_intron_selected_svr$nonsplice_variant_rate))
          
          Rsvrfpkm = get_rsquared_slope(prop.quantile=0.2,Xaxis=log10(by_intron_selected_svr[ by_intron_selected_svr$median_fpkm != 0 & !is.na(by_intron_selected_svr$median_fpkm),]$median_fpkm),
                                        Yaxis=by_intron_selected_svr[ by_intron_selected_svr$median_fpkm != 0 & !is.na(by_intron_selected_svr$median_fpkm),]$splice_variant_rate)
          
          Rnsvrfpkm = get_rsquared_slope(prop.quantile=0.2,Xaxis=log10(by_intron_selected_svr[ by_intron_selected_svr$median_fpkm != 0 & !is.na(by_intron_selected_svr$median_fpkm),]$median_fpkm),
                                         Yaxis=by_intron_selected_svr[ by_intron_selected_svr$median_fpkm != 0 & !is.na(by_intron_selected_svr$median_fpkm),]$nonsplice_variant_rate)
          
          data_summary = add_charac(data_summary,paste("rsquared_svr_fpkm;svr_class_",svr_class,";buscodataset_",busco_group,";quant",sep=""),"",Rsvrfpkm[1]  )
          data_summary = add_charac(data_summary,paste("slope_svr_fpkm;svr_class_",svr_class,";buscodataset_",busco_group,";quant",sep=""),"",Rsvrfpkm[2]  )
          data_summary = add_charac(data_summary,paste("rsquared_nsvr_fpkm;svr_class_",svr_class,";buscodataset_",busco_group,";quant",sep=""),"",Rnsvrfpkm[1]  )
          data_summary = add_charac(data_summary,paste("slope_nsvr_fpkm;svr_class_",svr_class,";buscodataset_",busco_group,";quant",sep=""),"",Rnsvrfpkm[2]  )
        }
      }
    }
  }
  data_summary$species = species
  
  all_dt = rbind(all_dt,data_summary)
}
